### PR TITLE
Adding an Error object

### PR DESCRIPTION
**Pull request by Nyholm**
The end goal of this PR is to be able to run the parser and output a well formatted error message. I would really like to be able to create good "github actions code comments" (or [commands](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message)). That is why I need to be able to get the errors in a more structured way. 

This PR does not break BC.

### CODE REVIEW

**Comment by Nyholm** [comment]
Oh.. my normal style of coding is to write a bunch of changes and then see if the CI is green. I guess Github's new "security feature" requires a maintainer's approval to run the CI on this PR.

File: lib/ErrorManager.php
Comment (wouterj) [inline lib/ErrorManager.php]: If we want to make an abstraction of the error handling in this library (which I agree with), I would say that we also move this code into an `StdErrorReporter` or the like (so that we can also e.g. add an `GitHubErrorReporter`).

File: lib/Error.php
Comment (greg0ire) [inline lib/Error.php]:
```suggestion
final class Error
```

File: lib/Error.php
Comment (greg0ire) [inline lib/Error.php]: Did you really mean to name this variable that way? :smile:

File: lib/ErrorManager.php
Comment (greg0ire) [inline lib/ErrorManager.php]:
```suggestion
            echo '⚠️ ' . $error->__toString() . "\n";
```

File: lib/Error.php
Comment (greg0ire) [inline lib/Error.php]: This does not seem to be used the magic way, so let's rename it?
```suggestion
    public function asString(): string
```

File: lib/ErrorManager.php
Comment (greg0ire) [inline lib/ErrorManager.php]: Might be more handy to have a method that takes scalar arguments and builds an `Error` VO internally, instead of having `addError(new Error(…))` every time

File: lib/Error.php
Comment (Nyholm) [reply lib/Error.php]: No, My energy level was super low when I wrote this.

File: lib/ErrorManager.php
Comment (Nyholm) [reply lib/ErrorManager.php]: Sure

File: lib/ErrorManager.php
Comment (Nyholm) [reply lib/ErrorManager.php]: I pushed some changes without the abstraction. When we are adding a third format, then I think it starts to make sense with an abstraction. Right now it is just a simple if statement.

**Comment by Nyholm** [comment]
Hm.. I need approval again..

**Comment by greg0ire** [comment]
Consider making a PR that upgrades to `doctrine/coding-standard 9` or any other small thing, that way I can approve and merge it and you will be autonomous on this one ;)

**Comment by Nyholm** [comment]
I am a doctrine org contributor already... 

Hm.. anyways: https://github.com/doctrine/rst-parser/pull/149

File: lib/ErrorManager.php
Comment (Nyholm) [reply lib/ErrorManager.php]: With this change, I wonder if I should use the existing `error`/`warning` method instead. The only downside is the argument order.. But I can live with `->error($message, $throwable, $file, $line)`

File: lib/Environment.php
Comment (greg0ire) [inline lib/Environment.php]: This is not valid for old versions of PHP.
```suggestion
            $this->getCurrentFileName()
```

Comment (Nyholm) [reply lib/Environment.php]: Thank you. I could not find that syntax error.

File: lib/ErrorManager.php
Comment (greg0ire) [reply lib/ErrorManager.php]: Or you could target 0.5.x and break BC :)

Comment (Nyholm) [reply lib/ErrorManager.php]: Which way will make it quicker for me to use this feature in my project?
 =)

Comment (greg0ire) [reply lib/ErrorManager.php]: Well we just released `0.4.0`, so it wouldn't make a lot of sense to release `0.5.0` with just that change. That being said, since `ErrorManager` is not final, adding arguments such as `$file` and `$line` would be a BC break for extending classes, so I'm not sure there is a choice either way.

Comment (greg0ire) [reply lib/ErrorManager.php]: I just merged https://github.com/doctrine/rst-parser/pull/144 that will make 0.5.0 less empty, and also make it less likely for such changes to become BC breaks.

**Comment by Nyholm** [comment]
Could I ask you to do a 0.4.x -> 0.5.x merge?

**Comment by Nyholm** [comment]
I've updated the PR to target 0.5.x. CI will be green after 0.4.x -> 0.5.x merge.

**Comment by greg0ire** [comment]
I'm AFK until tonight, will try to!

**Review by greg0ire — APPROVED** [review]

**Comment by Nyholm** [comment]
Thank you for the help. 

Im done with this PR now.

**Review by wouterj — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @Nyholm !

**Comment by Nyholm** [comment]
Wohoo. Thank you for the reviews and merge!

**Pull request merged**
